### ISSUE TITLE
Unable to disable zincati.service using Ignition

### ISSUE THREAD

**Issue by dustymabe** [coreos/fedora-coreos-tracker#392]
As reported by a user in [a discussion forum thread](https://discussion.fedoraproject.org/t/how-do-we-kick-off-manual-updates/17089/12?u=dustymabe) for some reason disabling `zincati.service` via Ignition doesn't seem to be working right. There is probably a corner case in here that isn't being accounted for somehow.

Here is my fcc:

```
variant: fcos
version: 1.0.0
systemd:
  units:
    - name: serial-getty@ttyS0.service
      dropins:
      - name: autologin-core.conf
        contents: |
          [Service]
          # Override Execstart in main unit
          ExecStart=
          # Add new Execstart with `-` prefix to ignore failure
          ExecStart=-/usr/sbin/agetty --autologin core --noclear %I $TERM
          TTYVTDisallocate=no
    - name: zincati.service
      enabled: false
```

Here is what I see on the booted system:

```
[core@localhost ~]$ systemctl is-enabled zincati
enabled
[core@localhost ~]$ systemctl is-active zincati
active
[core@localhost ~]$ 
[core@localhost ~]$ grep zincati /etc/systemd/system-preset/20-ignition.preset
disable zincati.service
[core@localhost ~]$ grep zincati /usr/lib/systemd/system-preset/40-coreos.preset
enable zincati.service
```

So the `disable` preset did get put in `/etc/` appropriately but for some reason it wasn't applied properly. If I ask systemd to re-evaluate it seems to do the right thing:

```
[core@localhost ~]$ sudo systemctl preset zincati.service
Removed /etc/systemd/system/multi-user.target.wants/zincati.service.
[core@localhost ~]$ systemctl is-enabled zincati
disabled
```

**Version Info**:

```
[core@localhost ~]$ rpm-ostree status
State: idle
AutomaticUpdates: disabled
Deployments:
● ostree://fedora:fedora/x86_64/coreos/testing
                   Version: 31.20200210.2.0 (2020-02-10T21:52:48Z)
                    Commit: 3e4a6a48ed8d6817081c902bc2aa2bbe9df3302e659d4f42d933f9abb22914e8
              GPGSignature: Valid signature by 7D22D5867F2A4236474BF7B850CB390B3C3359C4
[core@localhost ~]$ rpm -q ignition systemd
ignition-2.1.1-5.git40c0b57.fc31.x86_64
systemd-243.6-1.fc31.x86_64
```

**Comment by dustymabe** [comment]
Looks like this and #229 may be the same problem

**Comment by dustymabe** [comment]
ok here may be a clue:

```
[   15.943369] systemd[1]: Running with unpopulated /etc.

Welcome to Fedora CoreOS 31.20200319.dev.3!

[   15.949061] systemd[1]: Initializing machine ID from KVM UUID.
[   16.244489] systemd[1]: Populated /etc with preset unit settings.
```

so when systemd first starts `/etc/` is empty so it can't see what units are or are not enabled

**Comment by dustymabe** [comment]
well I guess it depends on what is really meant by `Running with unpopulated /etc`

**Comment by jlebon** [comment]
Re. presets, a lot of discussions here: https://github.com/coreos/fedora-coreos-config/pull/77. That said, the `enabled: false` from Ignition should result in a later `disable zincati` preset entry which wins over the default. (Which yeah, this approach runs against https://github.com/coreos/ignition/issues/588). Hmm, so something funny is going on here.

**Comment by dustymabe** [comment]
if I use `rd.break=pre-pivot` on the kernel command line I see the presets file:

```
pre-pivot:/# cat /sysroot/etc/systemd/system-preset/20-ignition.preset 
disable zincati.service
```

and running `preset-all` against the other root does the right thing:

```
pre-pivot:/# systemctl --root=/sysroot/ preset-all
Removed /sysroot/etc/systemd/system/multi-user.target.wants/zincati.service.
```

**Comment by dustymabe** [comment]
wow, thanks @jlebon - that was certainly not what I expected to be the problem.

**Comment by cgwalters** [comment]
A workaround until the systemd PR lands is probably something like adding a fragment with `ConditionPathExists=/enoent` or whatever instead of directly disabling.

**Comment by ashcrow** [comment]
Related: https://bugzilla.redhat.com/show_bug.cgi?id=1822865

**Comment by sgohl** [comment]
> A workaround until the systemd PR lands is probably something like adding a fragment with ConditionPathExists=/enoent or whatever instead of directly disabling.

how would a butane config unit look like with this approach?
My main problem is that my provisioning script collides with zincati:

```
May 26 21:06:06 rdbla229 zincati[744]: [INFO ] target release '34.20210427.3.0' selected, proceeding to stage it
May 26 21:06:07 rdbla229 zincati[744]: [ERROR] failed to stage deployment: rpm-ostree deploy failed:
May 26 21:06:07 rdbla229 zincati[744]:     error: Transaction in progress: --idempotent install open-vm-tools htop cronie
```
and it will not recover for a long time, showing:
```
May 26 21:14:26 rdbla229 zincati[732]: [ERROR] failed to register as driver: rpm-ostree deploy --register-driver failed:
May 26 21:14:26 rdbla229 zincati[732]:     error: Transaction in progress: --idempotent install open-vm-tools htop cronie
May 26 21:14:26 rdbla229 zincati[732]:      You can cancel the current transaction with `rpm-ostree cancel`
May 26 21:14:26 rdbla229 zincati[732]:
May 26 21:14:27 rdbla229 zincati[732]: [INFO ] initialization complete, auto-updates logic enabled
May 26 21:14:27 rdbla229 zincati[732]: [INFO ] update strategy: immediate
May 26 21:14:27 rdbla229 systemd[1]: Started Zincati Update Agent.
May 26 21:14:27 rdbla229 zincati[732]: [INFO ] current release detected as not a dead-end
```

But I want to have an updated system quite after provisioning. I think your approach would fit my need to enable zincati only after my provision script is done

**Comment by bgilbert** [comment]
@sgohl Something like this:

```yaml
variant: fcos
version: 1.0.0
systemd:
  units:
    - name: zincati.service
      dropins:
        - name: disable-zincati.conf
          contents: |
            [Unit]
            ConditionPathExists=/enoent
```

**Comment by jlebon** [comment]
> But I want to have an updated system quite after provisioning. I think your approach would fit my need to enable zincati only after my provision script is done

Assuming your provisioning script reboots, another approach is to run it with `Before=zincati.service`. That'll save you from having to then re-enable Zincati on reboot. You'll also want `After=systemd-machine-id-commit.service`.

**Comment by sgohl** [comment]
> You'll also want After=systemd-machine-id-commit.service

do you mean this condition for zincati or my provision script?

btw, actually I want to set zincati to periodic updates, but I'd like to have an opportunity to trigger an immediate update like it was possible with CoreOS. I know this is against your design principle, but I think it would not collide with something else since I'd see it as a hook into the immediate-routine - I may be wrong. But the way it is now, I am forced to wait for an update until the next window defined which leaves a bad feeling setting this host into production $now

**Comment by lucab** [comment]
@sgohl There are ways for doing that today, for example a `strategy = immediate` fragment under `/run` [hierarchy](https://coreos.github.io/zincati/usage/configuration/) which goes away after the first reboot.

> Actually I want to set zincati to periodic updates, but I'd like to have an opportunity to trigger an immediate update
> [...]
> (I know, i don't mean the staging, actually the reboot)

I think there is space to grow a primitive in Zincati to "temporarily bypass strategy" interactively. I had a similar need once in order to bypass a  faulty FleetLock server. There are several nuances in designing this though. Can you please record your expected usecase/flow directly in a Zincati ticket? We can probably find a middle-ground solution which helps in several maintenance cases without becoming a fully-interactive tool.

**Comment by sgohl** [comment]
from my perspective, it does not need to be interactive at all. What's actually missing, is just something like

```
          [updates]
          strategy = "immediate"
          reboot = false
```

because the staging/deployment is already done, we'd just reboot like you was used to with yum-cron. 
a notification about a needed reboot would be nice.

Anyway, static periodic windows are way too simple and insecure (in terms of availability).
Not all machines are in a cluster and even those who are, are surely fault-tolerant, but still some minimal service interruption happens, even if it's just a pacemaker-vip-move and the possible maintenance windows may vary.

I'd happily contribute in any way helpful, but I actually just starting to understand the whole thing (fedora coreos is really pretty new for me)

**Comment by lucab** [comment]
It does not work that way, by design. An arbitrary reboot does not change the booted OS, even if there is already a new staged commit.

Again, this ticket is not the better place to discuss Zincati changes. We can and do add strategies to Zincati to better suit general needs (https://github.com/coreos/zincati/pull/540 is in flight right now and may even solve this), but we do need to record the full usecase in a dedicated ticket, to make sure we understand the context first.

**Comment by jlebon** [comment]
As mentioned in https://github.com/coreos/fedora-coreos-tracker/issues/857#issuecomment-859261196, another easy way to work around this is to use `mask: true`.

**Comment by dustymabe** [comment]
We discussed this in [the community meeting today](https://lists.fedoraproject.org/archives/list/coreos@lists.fedoraproject.org/message/4MTHVVBEFGW2NEICVJ7I36KCLGJGC5EX/). An email was sent earlier this week as a last ditch effort to try to get feedback on https://github.com/systemd/systemd/pull/15205. We discussed alternatives in case that doesn't land and will work on those if the PR doesn't land.

**Comment by dustymabe** [comment]
We discussed this in [the community meeting today](https://lists.fedoraproject.org/archives/list/coreos@lists.fedoraproject.org/message/LOOOQUFHILOCHPXA2O54NRF7BWPOZMFH/).

```
13:16:29    dustymabe | #info there was no response to last week's request for
                      | feedback on https://github.com/systemd/systemd/pull/15205
                      | so we'll pursue a workaround solution.
```

```
13:25:46    dustymabe | #action jlebon bgilbert travier to discuss potential
                      | solutions for #392 and update the ticket and present the
                      | outcome at a future meeting.
```

**Comment by travier** [comment]
https://github.com/coreos/fedora-coreos-config/pull/363#issuecomment-1069323399:
> Actually, I think another approach here is to tweak Ignition so that enabled: false will also make sure that no enablement symlink exists in the real root on top of emitting a disable in the presets.

With the systemd PR moving forward, we will likely want to do something like this to fix the immediate issue while waiting for the full fix from upstream.

**Comment by jlebon** [comment]
@bgilbert @travier and I met to discuss this:
- There appears to be some movement in [the upstream systemd PR](https://github.com/systemd/systemd/pull/15205), so this may still be the long-term solution.
- In the short-term, we will have Ignition delete any enablement symlinks for services in the Ignition config (in addition to emitting a preset directive as today). This will allow us to have it fixed without waiting for the systemd patch to propagate.
- We discussed having Ignition across the board always handle enablement symlinks instead of presets, but there are corner cases there that would need consideration, and we'd rather stay away from "smearing" Ignition logic into the real root (see https://github.com/coreos/ignition/issues/588 for some discussions there).
- Regardless, Ignition also will need to learn about enablement symlinks for user-scoped [user services](https://github.com/coreos/ignition/pull/1303), which don't support presets.

**Comment by dustymabe** [comment]
Might be good to have a f-c-c external kola test for this too.

**Comment by dustymabe** [comment]
This landed in ignition v2.14.0.

**Comment by dustymabe** [comment]
The fix for this went into `stable` stream release `36.20220522.3.0`.
